### PR TITLE
Integration tests with a read only user

### PR DESCRIPTION
**Pull request by slvrtrn**
As expected: we cannot use response http compression for a read-only user

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -185,10 +185,12 @@ function formatSelectQuery(query: string, format: DataFormat): string {
...
 
-// it is a duplicate of `formatSelectQuery`, but it might differ in the future
-function formatCommandQuery(query: string, format: DataFormat): string {
+function formatCommandQuery(query: string, format: DataFormat | null): string {
   query = query.trim()
-  return query + ' \nFORMAT ' + format
+  if (format !== null) {
+    return query + ' \nFORMAT ' + format
+  }
+  return query
```
Comment (slvrtrn) [inline src/client.ts:197]: Some commands such as GRANT do not support the FORMAT clause, so I needed some way to disable `FORMAT ${format}` entirely. For now, it is an explicit null used as follows: 

```
await client.command({
  query,
  format: null
})
```
@mshustov wdyt?

File: src/client.ts [hunk 2]
```diff
@@ -139,15 +139,15 @@ export class ClickHouseClient {
...
+    const format = params.format === undefined ? 'JSON' : params.format
     const query = formatCommandQuery(params.query, format)
 
     const stream = await this.connection.command({
       query,
       ...this.getBaseParams(params),
     })
 
-    return new Rows(stream, format)
+    return new Rows(stream, 'JSON')
```
Comment (slvrtrn) [inline src/client.ts]: oops, will fix

File: src/client.ts (hunk 1 above)
Comment (mshustov) [reply src/client.ts:197]: Ok. Another option might be explicit `false`.

File: src/client.ts [hunk 3]
```diff
@@ -44,7 +44,7 @@ export interface SelectParams extends BaseParams {
 
 export interface CommandParams extends BaseParams {
   query: string
-  format?: DataFormat
+  format?: DataFormat | null
```
Comment (mshustov) [inline src/client.ts]: Could you add a comment explaining the purpose of `null`?

File: src/connection/adapter/http_search_params.ts [hunk 4]
```diff
@@ -35,7 +35,9 @@ export function toSearchParams({
 
   if (clickhouse_settings !== undefined) {
     for (const [key, value] of Object.entries(clickhouse_settings)) {
-      params.set(key, formatQuerySettings(value))
+      if (value !== undefined) {
```
Comment (mshustov) [inline src/connection/adapter/http_search_params.ts:38]: Could you add a unit test?

File: __tests__/integration/read_only_user.test.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            `To execute this query it's necessary to have grant INSERT`
+        ),
+      })
+    )
+  })
+
+  // TODO: enabled by default
+  //  we can try to restrict the access using https://clickhouse.com/docs/en/sql-reference/statements/create/row-policy/
+  //  as suggested in https://github.com/ClickHouse/ClickHouse/issues/2451#issuecomment-1162497824
+  it.skip('should fail to query system tables', async () => {
```
Comment (mshustov) [inline __tests__/integration/read_only_user.test.ts]: What prevents us from specifying a row policy and un-skipping the test? Sorry, it's not entirely clear from the comment.

File: __tests__/integration/read_only_user.test.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    await expect(
+      client.insert({
+        table: tableName,
+        values: [[43, 'foobar', [5, 25]]],
+      })
+    ).rejects.toEqual(
+      expect.objectContaining({
+        message: expect.stringContaining(
+          'Not enough privileges. ' +
+            `To execute this query it's necessary to have grant INSERT`
```
Comment (mshustov) [inline __tests__/integration/read_only_user.test.ts]: Optional: I'm on the fence with such assertions. The more details it is, the higher risk of flakiness due to some changes in string content. IMO a minimalistic assertion against `Not enough privileges` is okay.

File: __tests__/integration/read_only_user.test.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  beforeAll(async () => {
+    username = `clickhousejs__read_only_user_${guid()}`
+    const database = getTestDatabaseName()
+    const defaultClient = createTestClient()
+    const env = getClickHouseTestEnvironment()
+
+    // Bootstrap read only user
+    let createUser: string
+    let grant: string
+    switch (env) {
```
Comment (mshustov) [inline __tests__/integration/read_only_user.test.ts]: optional: I'd suggest extracting it in a test helper to reduce the cognitive load for a developer reading the test.

File: src/client.ts (hunk 1 above)
Comment (slvrtrn) [reply src/client.ts:197]: Yep, I like explicit `false` better

**Pull request merged**